### PR TITLE
Wildcard rework (backward incompatible)

### PR DESCRIPTION
**Pull request by akx**
This is a full rework of the wildcard system (now a subpackage :-) ).

# Goals

The main goal here is to

* make finding wildcard files faster (it certainly will be, since the hierarchy will be in-memory)
* make it possible to combine .txt wildcard files with noodle-soup-prompts "pantry" files, see https://github.com/adieyal/sd-dynamic-prompts/pull/231#issuecomment-1460693046
  * JSON and YAML format supported!

A future PR might also add a "find wildcard files with a given string" feature.

# Notes

* ⚠️ Since this renames a bunch of packages and removes APIs that are no longer necessary, this is backward incompatible with old versions.
  * I think we'd want to bump versions to 0.11 or even 0.20 :)
  * The `get_collections()` API that was really quite specific to SDDP is now gone! `Collection` is now the term used for a wildcard file (or a list of strings when read from a structured file) 
* The entire tree structure is now cached; in downstream uses such as sd-dynamic-prompts, it might be a good idea to call `wildcard_manager.clear_cache()` before a generation batch (but not before each generation).
* We also now support a single JSON or YAML file hosting an entire _tree_ of wildcards; this could make `collections` support much nicer...

### CODE REVIEW

**Comment by adieyal** [comment]
Two questions:

1. Are you considering full-text search? - I might have a large wildcard collection and may not know whether certain wildcards exist and where to find them.
2. What are your thoughts on using wildcards to create re-usable subprompts (see https://github.com/adieyal/sd-dynamic-prompts/discussions/325). The mechanism is similar, but the concept is slightly different. This is currently possible, but since newlines are significant in wildcard files, it is impossible to keep the formatting of detailed sub-prompts which makes them hard to read.

**Comment by akx** [comment]
> 1. Are you considering full-text search? - I might have a large wildcard collection and may not know whether certain wildcards exist and where to find them.

Yep, I have the same problem, and that could well be implemented later on top of this. :)

> 2. [...] Since newlines are significant in wildcard files, it is impossible to keep the formatting of detailed sub-prompts which makes them hard to read.

This PR will eventually also enable different formats for wildcard files; after all, NSP pantry files are JSON of the form `{"wildcard": ["value", "value", "value"], ...}`, so a sufficiently advanced user could already use them for multi-line prompts.

Since all environments that would run Stable Diffusion will already need YAML reading support for model configs, "pantry-style" files in YAML format wouldn't be far-fetched, and then

```yaml
animal-prompts:
  - |
    very
    complex
    multi-line
    prompt
  - simple thing
  - blah blah
```
is a snap.

**Comment by akx** [comment]
Ah, actually, one more thing: I'll add support for nested collections in pantries...

File: src/dynamicprompts/wildcards/wildcard_manager.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    build_tree_from_path,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class WildcardManager:
+    def __init__(
+        self,
+        path: Path | None = None,
```
Comment (adieyal) [inline src/dynamicprompts/wildcards/wildcard_manager.py:22]: What are your thoughts on allowing a string path as well as a pathlib Path?

**Comment by adieyal** [comment]
Overall, looks great. v0.2 sounds good, even v1.0. I'm pretty confident that the library is mature enough.

File: src/dynamicprompts/wildcards/wildcard_manager.py (hunk 1 above)
Comment (akx) [reply src/dynamicprompts/wildcards/wildcard_manager.py:22]: Sure, that makes sense. It'd be just one cast within the ctor anyway. :)

**Comment by akx** [comment]
> Overall, looks great. v0.2 sounds good, even v1.0. I'm pretty confident that the library is mature enough.

I wouldn't go to 1.0 just yet TBH - but 0.20, not 0.2 (since we're at 0.10 right now and 2 < 10 :) )

I'd also like to see how this API works out for sd-dynamic-prompts before we cut a new version (since this really is backward incompatible, and I may have well missed something). To that end, I'll make this a draft again for now and look at that soon (tomorrow most likely, tonight's movie night 🍿)

**Comment by adieyal** [comment]
> I wouldn't go to 1.0 just yet TBH 
What do you see 1.0  that we don't currently have? Of course there are always features to add but I don't think anything is missing.

 - but 0.20, not 0.2 (since we're at 0.10 right now and 2 < 10 :) )
 obvs - typo :)
> 
> I'd also like to see how this API works out for sd-dynamic-prompts before we cut a new version (since this really is backward incompatible, and I may have well missed something). To that end, I'll make this a draft again for now and look at that soon (tomorrow most likely, tonight's movie night popcorn)

Enjoy!

**Comment by akx** [comment]
@adieyal It was just as well that I did the work for SDDP too – I quite significantly reworked this PR too when I realized some awkwardnesses...

Notably:

* `WildcardFile` is now named `WildcardCollection` because it felt silly to have a `ListWildcardFile` that wasn't an on-disk file at all
* on that note, `collections` support is gone from here; it's replaced by a single function in https://github.com/adieyal/sd-dynamic-prompts/pull/341

**Review by adieyal — APPROVED** [review]
I like the refactor. Some thoughts about future work:

1. Currently wildcards are strings. There have been requests for weights as well as matching negative prompts, i.e. negative wildcards that should be used in conjunction with positive wildcards. I haven't felt a strong need for this, but was wondering what you thought about allowing for this sort of additional metadata.
2. I think that a significant contribution that we could make with dynamic prompts is provide a mechanism for sharing subprompts/prompt snippets. For example, in this [Reddit post](https://www.reddit.com/r/StableDiffusion/comments/10xqwjm/i_come_bearing_gifts_infinite_random_rpg/), the OP provides a pretty extensive DP prompt for generating  RPG characters. [Public Prompts](https://publicprompts.art/) also provides interesting prompt templates for re-use. It would be great to help facilitate something like that, perhaps through the pantry mechansim. Do you think this idea fits in the wildcard paradigm, or is it a new concept to be fleshed out?

**Comment by akx** [comment]
> Currently wildcards are strings. [...] I haven't felt a strong need for this, but was wondering what you thought about allowing for this sort of additional metadata.

I'm all up for it. I was almost adding a `WildcardEntry` dataclass here already (e.g. for `weight`, see https://github.com/adieyal/sd-dynamic-prompts/issues/317), but then decided for a less intrusive PR for the time being. 

> There have been requests for weights as well as matching negative prompts, i.e. negative wildcards that should be used in conjunction with positive wildcards

So you mean that choosing a positive wildcard would somehow automagically choose the equivalent negative wildcard? That'll naturally need some cooperation in SDDP, but... sounds doable...

> I think that a significant contribution that we could make with dynamic prompts is provide a mechanism for sharing subprompts/prompt snippets. It would be great to help facilitate something like that, perhaps through the pantry mechansim. Do you think this idea fits in the wildcard paradigm, or is it a new concept to be fleshed out?

That's absolutely a valid use case for wildcards (even if the name "wildcard" ends up being a bit of a misnomer then, but as we know, naming things is hard etc.). 

In fact, something that could maybe be pretty cool to implement is to gather data in e.g. a collaborative Google Sheet (because everyone knows how to use those), and transform it into pantry files (or just a synthetic `WildcardTree`).

Also now that it's pretty easy to merge WildcardTrees, maybe we should just start shipping an always-on curated set of wildcards in SDDP, too? E.g. the above Reddit prompt could be a `__dp/skud14-rpg-character__`.

**Pull request merged**